### PR TITLE
Ravendb 5055 : index merger

### PR DESCRIPTION
**Pull request by aviv86**

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+            _queryProcessed = true;
+
+            foreach (var initializer in objectCreateExpression.Initializers)
+            {
+                CollecetFieldNamesAndSelectsFromMemberDeclarator(initializer);
+            }
+        }
+
+        private void CollecetFieldNamesAndSelectsFromMemberDeclarator(AnonymousObjectMemberDeclaratorSyntax expression)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs:79]: typo: Collecet

File: src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        }
+
+        public override void VisitInvocationExpression(InvocationExpressionSyntax invocationExpression)
+        {
+            var memberReferenceExpression = invocationExpression.Expression as MemberAccessExpressionSyntax;
+
+            if (memberReferenceExpression == null)
+            {
+                base.VisitInvocationExpression(invocationExpression);
+            }
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs:106]: Shouldn't we `return` here?

File: src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                    break;
+                default:
+                    base.VisitInvocationExpression(invocationExpression);
+                    break;
+            }
+
+            if (lambdaExpression == null)
+            {
+                base.VisitInvocationExpression(invocationExpression);
+            }
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs:129]: Also I think we could `return` here or simplify this piece of code to:

```
if (lambdaExpression != null)
   ProcessQuery(lambdaExpression.Body);

base.VisitInvocationExpression(invocationExpression);
```

File: src/Raven.Server/Documents/Indexes/IndexMerging/IndexData.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                var assignmentExpression = SyntaxFactory.AnonymousObjectMemberDeclarator(name, curExpr.Value);
+                memberDeclarators = memberDeclarators.Add(assignmentExpression);
+            }
+
+            anonymousObjectCreationExpression = SyntaxFactory.AnonymousObjectCreationExpression(memberDeclarators);
+
+            if (InvocationExpression != null)
+            {
+                var memberAccessExpression = InvocationExpression.Expression as MemberAccessExpressionSyntax;
+                var invocExp = SyntaxFactory.InvocationExpression(memberAccessExpression)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/IndexMerging/IndexData.cs:59]: Can `memberAccessExpression` be null?

Comment (aviv86) [reply src/Raven.Server/Documents/Indexes/IndexMerging/IndexData.cs:59]: yes i think you're right, InvocationExpression.Expression can be an IdentifierName and then memberAccessExpression is null. I added a check for that.
thanks

File: src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs (hunk 1 above)
Comment (aviv86) [reply src/Raven.Server/Documents/Indexes/IndexMerging/CaptureSelectNewFieldNamesVisitor.cs:79]: fixed, thanks

**Pull request merged**